components/feature/addons/src/main/java/mozilla/components/feature/addons/AddonManager.kt

```diff
@@ -49,6 +64,47 @@ class AddonManager(
             throw AddonManagerException(throwable)
         }
     }
+
+    /**
+     * Try updates the provided [extensionId] by updating it in the the [store] and re-attaching
+     * all the handlers.
+     */
+    fun tryUpdateExtension(extensionId: String, onFinish: ((Status) -> Unit)) {
+        val extension = WebExtensionSupport.installedExtensions[extensionId]
+
+        if (extension == null) {
+            onFinish(Status.NotInstalled)
+            return
+        }
+
+        val onSuccess: ((WebExtension?) -> Unit) = { updatedExtension ->
+            val status = if (updatedExtension == null) {
+                Status.NoNewAvailableUpdate
+            } else {
+                WebExtensionSupport.markExtensionAsUpdated(store, updatedExtension)
+                Status.SuccessfullyUpdated
+            }
+            onFinish(status)
+        }
+
+        val onError: ((String, Throwable) -> Unit) = { message, exception ->
+            onFinish(Status.Error(message, exception))
+        }
+
+        engine.updateWebExtension(extension, onSuccess, onError)
+    }
+
+    @Suppress("EXPERIMENTAL_API_USAGE")
+    private fun registerNewExtensionsForUpdates() {
+        store.flowScoped { flow ->
+            flow.ifChanged { it.extensions }
```

Comment:
@csadilek @psymoon  would you mien giving me a hand? I'm trying to observe only new extensions when they are installed to register them for checking for updates. But I end up getting the whole collection the ones that hasn't change and the new ones.

Reply:
@Amejia481 yes, that's by design, the way this API behaves is that you always first get the current state (change from no state observed -> current state), and then only changes (last observed state -> changed state).

Reply:
@Amejia481 [`Flow.distinctUntilChanged()`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/distinct-until-changed.html) may be useful. You may need a mediator observable generally to map/collect the latest data set (speaking from Rx perspective)

Reply:
Awesome, let me take a look at it, thanks!

Reply:
@Amejia481 @psymoon I don't think this is needed here. We know all existing addons here on startup and can register them for updates. Installing a new addon will also go through the `AddonManager` and we can register for updates then i.e. we're about to add an `installAddon` call to the manager. So, we don't actually need to go through the store here.

Reply:
In other words `addons` can't change without the `AddonManager` triggering it. So, I can just call `registerForFutureUpdates`, as needed.

Reply:
Does that make sense? We can pair up on that too.

Reply:
Go it 👍 !

Reply:
I will summit the pr with the `registerForFutureUpdates` function to be integrated in the follow up work.

Reply:
👍